File: 21. Merge Two Sorted Lists/21. Merge Two Sorted Lists.py
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # base cases
        if not list1:
            return list2
        if not list2:
            return list1
        
        # determine the head of the merged list
        if list1.val <= list2.val:
            head = list1
            list1 = list1.next
        else:
            head = list2
            list2 = list2.next
        
        # pointer to traverse the merged list
        current = head
        
        # merge the lists
        while list1 and list2:
            if list1.val <= list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next
            current = current.next
        
        # append the remaining nodes of list1 or list2
        if list1:
            current.next = list1
        else:
            current.next = list2
        
        return head

File: 21. Merge Two Sorted Lists/21. Merge Two Sorted Lists.go
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    // Обработка базовых случаев
    if list1 == nil {
        return list2
    }
    if list2 == nil {
        return list1
    }
    
    // Определение головы объединенного списка
    var head *ListNode
    if list1.Val <= list2.Val {
        head = list1
        list1 = list1.Next
    } else {
        head = list2
        list2 = list2.Next
    }
    
    // Указатель для обхода объединенного списка
    current := head
    
    // Слияние списков
    for list1 != nil && list2 != nil {
        if list1.Val <= list2.Val {
            current.Next = list1
            list1 = list1.Next
        } else {
            current.Next = list2
            list2 = list2.Next
        }
        current = current.Next
    }
    
    // Добавление оставшихся узлов list1 или list2
    if list1 != nil {
        current.Next = list1
    } else {
        current.Next = list2
    }
    
    return head
}
